core/net/mac/contikimac/contikimac-async.c

```diff
@@ -0,0 +1,1438 @@
+#include "lib/list.h"
+#include "sys/pt.h"
+#include "sys/rtimer.h"
+#include "sys/compower.h"
+#include "sys/isr-control.h"
+
+#include <stdio.h>
+#include <stdbool.h>
+#include <stdint.h>
+/*---------------------------------------------------------------------------*/
+/*------------------------------CONFIG CHECKING------------------------------*/
+/*---------------------------------------------------------------------------*/
+#if !(RTIMER_MULTIPLE_ACCESS)
+#error "contikimac-async requires multiple rtimer support!"
+#endif
+/*---------------------------------------------------------------------------*/
+/*-----------------------------------TYPES-----------------------------------*/
+/*---------------------------------------------------------------------------*/
+enum cm_event {
+  CM_EVENT_TX_DONE_OK,
+  CM_EVENT_TX_DONE_ERR,
+  CM_EVENT_TX_COLLISION,
+  CM_EVENT_CCA_DONE_BUSY,
+  CM_EVENT_CCA_DONE_IDLE,
+  CM_EVENT_CCA_DONE_ERR,
+  CM_EVENT_RX_BURST_DONE,
+  CM_EVENT_RX_ARRIVED,
+  CM_EVENT_GOT_ACK,
+  CM_EVENT_TIME_ERROR,
+  CM_EVENT_TIME_EXPIRED,
+  CM_EVENT_ON_OK,
+  CM_EVENT_ON_ERR,
+  CM_EVENT_OFF_OK,
+  CM_EVENT_OFF_ERR,
+  CM_EVENT_NONE
+};
+
+enum tx_state {
+  TX_STATE_IDLE,
+  TX_STATE_PHASE_PENDING,
+  TX_STATE_PHASE_WAIT,
+  TX_STATE_PENDING,
+  TX_STATE_RUNNING,
+  TX_STATE_ACK_WAIT,
+  TX_STATE_NO_ACK,
+  TX_STATE_GOT_ACK,
+  TX_STATE_DONE_OK,
+  TX_STATE_DONE_COLLISION,
+  TX_STATE_DONE_ERR,
+  TX_STATE_DONE_NOACK
+};
+
+enum rx_state {
+  RX_STATE_CYCLE_WAIT,
+  RX_STATE_CCA,
+  RX_STATE_PACKET_SEEN,
+  RX_STATE_PACKET_GOT,
+  RX_STATE_PACKET_BURST
+};
+
+struct mac_info {
+  /* mac callback data */
+  mac_callback_t mac_cb;
+  void *mac_ptr;
+
+  /* List of queued packets */
+  struct rdc_buf_list *buf_list;
+};
+/*---------------------------------------------------------------------------*/
+/*----------------------------------DEFINES----------------------------------*/
+/*---------------------------------------------------------------------------*/
+#define DEBUG 0
+#if DEBUG
+#include <stdio.h>
+#define PRINTF(...) printf(__VA_ARGS__)
+#define PRINTDEBUG(...) printf(__VA_ARGS__)
+#else
+#define PRINTF(...)
+#define PRINTDEBUG(...)
+#endif
+
+#define USE_RX_CCA_AS_TX_CCA 1
+
+#define TX_FLAG_BROADCAST     0x01
+#define TX_FLAG_RCVR_KNOWN    0x02
+#define TX_FLAG_RCVR_AWAKE    0x04
+#define TX_FLAG_PENDING       0x08
+#define TX_FLAG_CHANNEL_CLEAR 0x10
+
+#define LOCK_UNLOCKED 0
+#define LOCK_TX_OWNED 1
+#define LOCK_RX_OWNED 2
+
+#define LOCK_ACQUIRE_SUCCESS 0
+#define LOCK_ACQUIRE_FAILURE 1
+
+#define CYCLE_MIN_SETUP RTIMER_MINIMAL_SAFE_SCHEDULE
+
+#define RTIMER_INTER_PACKET_DEADLINE (RTIMER_ARCH_SECOND / 32)
+
+#define YIELD_ON_OR_GOTO(pt, stmnt, label) \
+  PT_YIELD_ON(pt, if((stmnt)) { goto label; })
+#define STROBE_YIELD_ON_OR_GOTO(stmnt, label) \
+  YIELD_ON_OR_GOTO(&strobe_state.pt, stmnt, label)
+#define CCA_YIELD_ON_OR_GOTO(stmnt, label) \
+  YIELD_ON_OR_GOTO(&cca_state.pt, stmnt, label)
```

Comment:
Nevermind.  I realize, now, that the functions, on which those macroes are used, return an error code (0 means OK, 1 means failure), not a boolean value.  So, those names are good.